libraries/eosiolib/core/eosio/crypto_bls_ext.hpp

```diff
@@ -157,5 +334,259 @@ namespace eosio {
+
+    // Construct an extensible-output function based on SHA256
+    void xmd_sh256(
+        char *buf,
+        int buf_len,
+        const char *in,
+        int in_len,
+        const char *dst,
+        int dst_len
+    ) {
+        const unsigned int SHA256HashSize = 32;
+        const unsigned int SHA256_Message_Block_Size = 64;
+        const unsigned ell = (buf_len + SHA256HashSize - 1) / SHA256HashSize;
+        if (buf_len < 0 || ell > 255 || dst_len > 255) {
+            return;
+        }
+        const uint8_t Z_pad[SHA256_Message_Block_Size] = { 0, };
+        const uint8_t l_i_b_0_str[] = {
+            static_cast<uint8_t>(buf_len >> 8),
+            static_cast<uint8_t>(buf_len & 0xff),
+            0,
+            static_cast<uint8_t>(dst_len)
+        };
+        const uint8_t *dstlen_str = l_i_b_0_str + 3;
+        uint8_t b_0[SHA256HashSize];
+        bls12_381::sha256 sha;
+        sha.update(Z_pad, SHA256_Message_Block_Size);
+        sha.update(in, in_len);
+        sha.update(l_i_b_0_str, 3);
+        sha.update(dst, dst_len);
+        sha.update(dstlen_str, 1);
+        sha.digest(b_0);
+        uint8_t b_i[SHA256HashSize + 1] = { 0, };
+        for (unsigned i = 1; i <= ell; ++i) {
+            for (unsigned j = 0; j < SHA256HashSize; ++j) {
+                b_i[j] = b_0[j] ^ b_i[j];
+            }
+            b_i[SHA256HashSize] = i;
+            bls12_381::sha256 s;
+            s.update(b_i, SHA256HashSize + 1);
+            s.update(dst, dst_len);
+            s.update(dstlen_str, 1);
+            s.digest(b_i);
+            const int rem_after = buf_len - i * SHA256HashSize;
+            const int copy_len = SHA256HashSize + (rem_after < 0 ? rem_after : 0);
+            memcpy(buf + (i - 1) * SHA256HashSize, b_i, copy_len);
+        }
+    }
+    bls_s scalar_fromBE(const bls_s& in) {
+        std::array<uint64_t, 8> out;
+        for(uint64_t i = 0; i < 8; i++) {
+            uint64_t temp;
+            memcpy(&temp, &in[in.size() - i*8 - 8], sizeof(uint64_t));
+            out[i] = htobe64(temp);
+        }
+        return reinterpret_cast<bls_s&&>(std::move(out));
+    }
+    void g2_fromMessage(const bls_g1_affine& msg, const std::string& dst, bls_g2& res) {
+
+        std::array<bls_s, 4> buf;
+        xmd_sh256(buf.data()->data(), sizeof(buf), msg.data(), msg.size(), dst.data(), dst.length());
+
+        bls_s k;
+        bls_fp2 t;
+        bls_g2 p, q;
+
+        k = scalar_fromBE(buf[0]);
+        bls_fp_mod(k, t[0]);
+        k = scalar_fromBE(buf[1]);
+        bls_fp_mod(k, t[1]);
+
+        bls_g2_map(t, p);
+
+        k = scalar_fromBE(buf[2]);
+        bls_fp_mod(k, t[0]);
+        k = scalar_fromBE(buf[3]);
+        bls_fp_mod(k, t[1]);
+
+        bls_g2_map(t, q);
+        bls_g2_add(p, q, res);
+    }
+
+    // pubkey and signature are assumed to be in RAW affine little-endian bytes
```

Comment:
Can you assert if they are not in the expected format?

Reply:
I don't think this is possible. this is just bytes where every random ill-formed public key can have corresponding private key. If you have any ideas how to facilitate it, please let me know.
this is possible with base64 but I don't know any way this is feasible in raw format